extensions/amp-date-countdown/1.0/date-countdown.js

```diff
@@ -0,0 +1,247 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {getLocaleStrings} from './messages';
+import {useAmpContext} from '../../../src/preact/context';
+import {useEffect, useMemo, useRef, useState} from '../../../src/preact';
+import {useResourcesNotify} from '../../../src/preact/utils';
+
+const NAME = 'DateCountdown';
+
+// Constants
+/** @const {number} */
+const MILLISECONDS_IN_DAY = 24 * 60 * 60 * 1000;
+
+/** @const {number} */
+const MILLISECONDS_IN_HOUR = 60 * 60 * 1000;
+
+/** @const {number} */
+const MILLISECONDS_IN_MINUTE = 60 * 1000;
+
+/** @const {number} */
+const MILLISECONDS_IN_SECOND = 1000;
+
+/** @const {number} */
+const DELAY = 1000;
+
+/** @const {Object<string, number>} */
+const TimeUnit = {
+  DAYS: 1,
+  HOURS: 2,
+  MINUTES: 3,
+  SECONDS: 4,
+};
+
+// Default prop values
+const DEFAULT_OFFSET_SECONDS = 0;
+const DEFAULT_LOCALE = 'en';
+const DEFAULT_WHEN_ENDED = 'stop';
+const DEFAULT_BIGGEST_UNIT = 'DAYS';
+
+/**
+ * @param {!DateCountdownPropsDef} props
+ * @return {PreactDef.Renderable}
+ */
+export function DateCountdown({
+  endDate,
+  timeleftMs,
+  timestampMs,
+  timestampSeconds,
+  offsetSeconds = DEFAULT_OFFSET_SECONDS,
+  whenEnded = DEFAULT_WHEN_ENDED,
+  locale = DEFAULT_LOCALE,
+  biggestUnit = DEFAULT_BIGGEST_UNIT,
+  render,
+  children,
+  ...rest
+}) {
+  useResourcesNotify();
+  const {playable} = useAmpContext();
+  const epoch = useMemo(
+    () =>
+      new Date(
+        getEpoch(endDate, timeleftMs, timestampMs, timestampSeconds) +
+          offsetSeconds * DELAY
+      ),
+    [endDate, timeleftMs, timestampMs, timestampSeconds, offsetSeconds]
+  );
+  const [timeleft, setTimeleft] = useState(epoch - Date.now());
+  const localeStrings = useMemo(() => getLocaleWord(locale), [locale]);
+  const rootRef = useRef(null);
+
+  useEffect(() => {
+    if (!playable || !rootRef.current) {
+      return;
+    }
+    const win = rootRef.current.ownerDocument.defaultView;
+    const interval = win.setInterval(() => {
+      const newTimeleft = epoch - Date.now() + DELAY;
```

Comment:
Continuing the earlier thread.

> Since the setInterval's callback is not being called until 1000ms (1 delay) from now, setting the correct time to be displayed at that time. So it will start with the initial value being displayed, and calculate the new time to be 1000ms less than the initial value, and will display that in 1000ms.

I still don't understand :/

Lets say it is currenty `time0` and we set the epoch to `time0 + 5000`.
When the interval fires at time=time0+3000, timeleft should be 2000 right? Which would mean `epoch - Date.now()`

Reply:
Thanks Jake, after thinking about it more, I agree with your analysis.

I looked into it more though, and I think we actually still need the `DELAY`, but for a different reason.

![image](https://user-images.githubusercontent.com/25626770/92619478-8c851b00-f28f-11ea-8c39-541046302215.png)

So in the screenshot `t` is `timeleft` and `n` is `newTimeleft`.  After the first interval, we end up with 3919 timeleft.  Then adding in the delay it becomes 4919 (displayed).  Since the logic further down actually floors this to 4000, the countdown appears to count smoothly 5, 4, etc...  This works since the interval does not run the callback at exactly 1000ms later, it runs it slightly later than that.

If we remove the delay, then the countdown will go 5, 3... and skip a count in the beginning. :/. (it will take 3919 and floor to 3000)

Reply:
I still dont like this 🤔 . Hitting LGTM now to unblock, but will continue to think on how to improve it